File: comparisons/hash_files.js
const fs = require("fs");
const crypto = require("crypto");

async function fileHash(filename, algorithm = "sha256") {
  return new Promise((resolve, reject) => {
    let shasum = crypto.createHash(algorithm);
    try {
      let s = fs.ReadStream(filename)
      s.on("data", function (data) {
        shasum.update(data)
      })
      // making digest
      s.on("end", function () {
        const hash = shasum.digest("hex")
        return resolve(hash);
      })
    } catch (error) {
      return reject("calc fail");
    }
  });
}

async function hashNTimes(n) {
  function hashi(index) {
    return fileHash("./cat.jpg");
  }

  var ps = [];
  for (var i = 0; i < n; i++) {
    ps.push(hashi(i));
  }

  return Promise.all(ps);
}

const NS_PER_SEC = 1e9;
const time = process.hrtime();

hashNTimes(1000)
  .then(() => {
    const diff = process.hrtime(time);
    console.log(`Execution time = ${diff[0] * NS_PER_SEC + diff[1]} nanoseconds`);
  });

File: comparisons/hash_files.go
package main

import (
  "crypto/sha256"
  "encoding/hex"
  "fmt"
  "io"
  "os"
  "runtime"
  "sync"
  "time"
)

func fileHash(file string) (hash string, err error) {
  hasher := sha256.New()
  f, err := os.Open(file)
  if err != nil {
    return
  }
  defer f.Close()

  if _, err = io.Copy(hasher, f); err != nil {
    return
  }

  return hex.EncodeToString(hasher.Sum(nil)), nil
}

func hashNTimes(n int) {
  var wg sync.WaitGroup

  for i := 0; i < n; i++ {
    wg.Add(1)
    go func(wg *sync.WaitGroup, index int) {
      defer wg.Done()

      _, err := fileHash("./cat.jpg")
      if err != nil {
        return
      }
    }(&wg, i)
  }

  wg.Wait()
}

func main() {
  nprocs := runtime.GOMAXPROCS(0)
  fmt.Printf("n CPU = %d\n", nprocs)

  t0 := time.Now()
  hashNTimes(1000)
  elapsed := time.Since(t0)
  fmt.Printf("Execution time = %d nanoseconds\n", elapsed.Nanoseconds())
}
